### PR TITLE
Occlusion sensitivity

### PR DESCRIPTION
**Pull request by rijobro**
Calculate the occlusion sensitivity of an image. Block part of the image, and record how the probability of the inferred class changes. Then iteratively move the occluded section across the image to build up a sensitivity map.

Once merged, I will add a small example to the end of the [2D classification tutorial](https://github.com/Project-MONAI/tutorials/blob/master/2d_classification/mednist_tutorial.ipynb). 

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/metrics/occlusion_sensitivity.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    area. Therefore, in the output image, higher values mean there was a
+    greater the drop in certainty, indicating the occluded region was more
+    important in the decision process.
+
+    By occlusion sensitivity, we mean how the probability of a given
+    prediction changes as the occluded section of an image changes. This can
+    be useful to understand why a network is making certain decisions.
+
+    Args:
+        model: model to use for inference
```
Comment (wyli) [inline monai/metrics/occlusion_sensitivity.py:46]: could you specify the expected shape for the model input and output and the labels? if it supports 3D, would be great to have a test case

**Review by wyli — COMMENTED** [review]
thanks!  very interesting metrics, perhaps could include a paper reference in the docstring about the method?

File: monai/metrics/occlusion_sensitivity.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    for i in trange(image.numel()):
+        # Get corresponding ND index
+        idx = np.unravel_index(i, im_shape)
+        # Get min and max index of bounding box.
+        min_idx = [max(0, i - margin) for i in idx]
+        max_idx = [min(j, i + margin) for i, j in zip(idx, im_shape)]
+
+        # Clone and replace target area with `pad_val`
+        occlu_im = image.clone()
+        occlu_im[..., min_idx[0] : max_idx[0], min_idx[1] : max_idx[1], min_idx[2] : max_idx[2]] = pad_val
```
Comment (casperdcl) [inline monai/metrics/occlusion_sensitivity.py:98]: For arbitrary `ndim`

```suggestion
        occlu_im[(...,) + tuple(slice(i, j) for i, j in zip(min_idx, max_idx))] = pad_val
```

**Comment by rijobro** [comment]
@wyli Thanks for the review. I've updated the documentation with expected input and outputs and added a 3D example. I've also examples in a [PR](https://github.com/Project-MONAI/tutorials/pull/47) in the Tutorials repository.

Once the tests all pass, I'm happy for this to be merged. Do you think you could give it another look over?

File: monai/metrics/occlusion_sensitivity.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+    sensitivity_im = torch.empty(0, dtype=torch.float32, device=image.device)
+
+    # If no bounding box supplied, output shape is same as input shape.
+    # If bounding box is present, shape is max - min + 1
+    output_im_shape = im_shape if b_box is None else b_box_max - b_box_min + 1
+    num_required_predictions = np.prod(output_im_shape)
+
+    # Loop 1D over image
+    for i in trange(num_required_predictions):
```
Comment (wyli) [inline monai/metrics/occlusion_sensitivity.py:153]: looks like there's some memory issue with the Windows test, I don't know the root cause for now... is it possible that this for loop is actually accumulating the model variable gradients, using a lot of memory? adding some `detach()` commands here might help?

File: monai/metrics/occlusion_sensitivity.py (hunk 1 above)
Comment (wyli) [reply monai/metrics/occlusion_sensitivity.py]: or maybe it's GitHub's issue https://www.githubstatus.com/incidents/k46g5dn2c4lt ...

File: monai/metrics/occlusion_sensitivity.py (hunk 3 above)
Comment (rijobro) [reply monai/metrics/occlusion_sensitivity.py:153]: The Windows test was originally failing but I decreased batch size and that seems to be passing now [link](https://github.com/Project-MONAI/MONAI/pull/1099/checks?check_run_id=1283108651#step:9:2582). The current error on Windows looks unrelated. I'll relaunch and see what happens.

**Review by wyli — APPROVED** [review]
thanks!

**Pull request merged**